src/pybind/mgr/nfs/export.py

```diff
@@ -424,63 +398,204 @@ def _create_user_key(self, entity, path, fs_name, fs_ro):
+                'clients': clients
+            }
             export = Export.from_dict(ex_id, ex_dict)
-            export.fsal.cephx_key = key
-            self._save_export(export)
+            export.validate(self.mgr)
+            self._save_export(cluster_id, export)
             result = {
-                    "bind": pseudo_path,
-                    "fs": fs_name,
-                    "path": path,
-                    "cluster": cluster_id,
-                    "mode": access_type,
-                    }
+                "bind": pseudo_path,
+                "path": bucket,
+                "cluster": cluster_id,
+                "mode": access_type,
+                "squash": squash,
+            }
             return (0, json.dumps(result, indent=4), '')
         return 0, "", "Export already exists"
 
-    def update_export(self, new_export):
-        old_export, update_user_caps = self._validate_export(new_export)
-        if update_user_caps:
-            self._update_user_id(new_export['path'], new_export['access_type'],
-                                 new_export['fsal']['fs_name'], new_export['fsal']['user_id'])
-        new_export = Export.from_dict(new_export['export_id'], new_export)
-        new_export.fsal.cephx_key = old_export.fsal.cephx_key
-        self._update_export(new_export)
-        export_ls = self.exports[self.rados_namespace]
-        if old_export not in export_ls:
-            # This happens when export is fetched by ID
-            old_export = self._fetch_export(old_export.pseudo)
-        export_ls.remove(old_export)
+    def update_export_1(
+            self,
+            cluster_id: str,
+            new_export_dict: Dict,
+    ) -> Tuple[int, str, str]:
+        for k in ['cluster_id', 'path', 'pseudo']:
+            if k not in new_export_dict:
+                raise NFSInvalidOperation(f'Export missing required field {k}')
+        if new_export_dict['cluster_id'] not in available_clusters(self.mgr):
+            raise ClusterNotFound()
+
+        new_export_dict['path'] = self.format_path(new_export_dict['path'])
+        new_export_dict['pseudo'] = self.format_path(new_export_dict['pseudo'])
+
+        old_export = self._fetch_export(new_export_dict['cluster_id'],
+                                        new_export_dict['pseudo'])
+        if old_export:
+            # Check if export id matches
+            if new_export_dict.get('export_id'):
+                if old_export.export_id != new_export_dict.get('export_id'):
+                    raise NFSInvalidOperation('Export ID changed, Cannot update export')
+            else:
+                new_export_dict['export_id'] = old_export.export_id
+        elif new_export_dict.get('export_id'):
+            old_export = self._fetch_export_obj(cluster_id, new_export_dict['export_id'])
+            if old_export:
+                # re-fetch via old pseudo
+                old_export = self._fetch_export(cluster_id, old_export.pseudo)
```

Comment:
tbh I'm not quite sure; this was adapting the old workaround here:
```
        export_ls = self.exports[self.rados_namespace]
        if old_export not in export_ls:
            # This happens when export is fetched by ID
            old_export = self._fetch_export(old_export.pseudo)
```

Reply:
I think it's because if we load by export_id we create a new object, whereas if we fetch by pseudo we get a reference to the existing in-memory object (which is what we want).